plausible: init at 1.3.0

Ma27 opened the pull request:
###### Motivation for this change

[plausible](https://plausible.io/) is a simple and privacy-friendly alternative to Google Analytics based on Elixir.

cc @happysalada @lheckemann @kvz 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Ma27 committed:
yarn2nix: set `name` to `${pname}-${version}` by default

Ma27 committed:
plausible: init at 1.3.0

r-rmcgibbo commented:
Result of `nixpkgs-review pr 124055` at 9eb694c5 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>plausible</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `glibcLocales`.
Near pkgs/servers/web-apps/plausible/default.nix:1:47:
```
  |
1 | { lib, stdenv, beamPackages, fetchFromGitHub, glibcLocales, cacert
  |                                               ^
```
</li>
    <li>warning: unused-argument

Unused argument: `stdenv`.
Near pkgs/servers/web-apps/plausible/default.nix:1:8:
```
  |
1 | { lib, stdenv, beamPackages, fetchFromGitHub, glibcLocales, cacert
  |        ^
```
</li>
    <li>warning: unused-argument

Unused argument: `cacert`.
Near pkgs/servers/web-apps/plausible/default.nix:1:61:
```
  |
1 | { lib, stdenv, beamPackages, fetchFromGitHub, glibcLocales, cacert
  |                                                             ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/fixup-phase.md">fixup-phase</a>

`fixupPhase` should not be overridden, use `postFixup` instead.

Near pkgs/development/beam-modules/mix-release.nix:77:3:
```
   |
77 |   fixupPhase = ''
   |   ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

happysalada commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,270 @@
+          documentation = [ "https://plausible.io/docs/self-hosting" ];
+          wantedBy = [ "multi-user.target" ]
+            ++ optional cfg.database.clickhouse.setup "clickhouse.service"
+            ++ optional cfg.database.postgres.setup "postgresql.service";
```
How about adding a `requires` for the postgres service?

Ma27 replied:
Hmm, at first I'll have to re-read the `systemd.unit`-manpages, I always forget the (slight) differences. But good point, I'm not really sure right now if a `WantedBy` actually makes sense here!

happysalada commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,270 @@
+          wantedBy = [ "multi-user.target" ]
+            ++ optional cfg.database.clickhouse.setup "clickhouse.service"
+            ++ optional cfg.database.postgres.setup "postgresql.service";
+          after = optional cfg.database.postgres.setup "plausible-postgres.service";
```
What about the clickhouse service here, the service will start properly even if the clickhouse service is not available? Just asking.

happysalada commented on pkgs/servers/web-apps/plausible/default.nix:
```diff
@@ -0,0 +1,85 @@
+    sha256 = "03lm1f29gwwixnhgjish5bhi3m73qyp71ns2sczdnwnbhrw61zps";
+  };
+
+  mixDeps = beamPackages.fetchMixDeps {
```
Not sure how you feel about the FOD vs Nix dependencies thing.
mix2nix has just been added to nixpkgs, and the PR to enable mix2nix generated dependencies to be passed to mixRelease, should be merged during the course of next week.
I'm just mentioning this in case you are interested.

Ma27 replied:
> Not sure how you feel about the FOD vs Nix dependencies thing.

I dislike it, but I use it when it's more practical :sweat_smile: 

In this case I simply wasn't aware of `mix2nix` since I wrote most of it back in Februrary (where most of the current tooling didn't even exist yet AFAIU). What's your experience with it? If it doesn't create additional workload, I'm happy to change this :)

happysalada replied:
I started testing it a couple of weeks ago, so basically take my opinion with a grain of salt.
However so far, I like it. I'm actually interested in figuring out the rough edges, so if you are interested, I can help with any problem you encounter with it. I worked with the person who did the PR to understand properly was was happening and potential failures. I'm by no means a mix expert, but I'm fairly confident we could get this to work should you encounter any problem. So if you like the nix approach better, I would say go for it!
I tested on plausible and it does generate a 1000 lines deps.nix file though. I think those are pretty standard dependencies, so not a problem, but if you don't want to bother with this, it makes sense too.

happysalada replied:
Here is what it would look like
```
{ lib, beamPackages, overrides ? (x: y: {}) }:

let
	buildRebar3 = lib.makeOverridable beamPackages.buildRebar3;
	buildMix = lib.makeOverridable beamPackages.buildMix;
	buildErlangMk = lib.makeOverridable beamPackages.buildErlangMk;

	self = packages // (overrides self packages);

	packages = with beamPackages; with self; {
		ecto_sql = buildMix rec {
			name = "ecto_sql";
			version = "3.6.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0y871vj9s54dmg8snqhdcml6p7ccmd8n67mxzszvcy2r5lzmrwv6";
			};

			beamDeps = [ db_connection ecto postgrex telemetry ];
		};

		ink = buildMix rec {
			name = "ink";
			version = "1.2.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0wilhwd65x03fd6saq8arwdjg3mdq1ij8ljwgrb973k9lnh0qmr6";
			};

			beamDeps = [ jason ];
		};

		ssl_verify_fun = buildRebar3 rec {
			name = "ssl_verify_fun";
			version = "1.1.6";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1026l1z1jh25z8bfrhaw0ryk5gprhrpnirq877zqhg253x3x5c5x";
			};

			beamDeps = [];
		};

		oauther = buildMix rec {
			name = "oauther";
			version = "1.1.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "13kqwl35zard67qjpcn3d6yl6dl9fnwpxifdris1hcj540qg8x4k";
			};

			beamDeps = [];
		};

		phoenix_html = buildMix rec {
			name = "phoenix_html";
			version = "2.14.3";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "03z8r285znlg25yi47d4l59s7jq58y4dnhvbxgp16npkzykrgmpg";
			};

			beamDeps = [ plug ];
		};

		phoenix_ecto = buildMix rec {
			name = "phoenix_ecto";
			version = "4.2.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1h23fy3pylaszh3l2zafq1a7fjwlwb3yw7dy09p0mb4wi6p1p2j7";
			};

			beamDeps = [ ecto phoenix_html plug ];
		};

		phoenix_pubsub = buildMix rec {
			name = "phoenix_pubsub";
			version = "2.0.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0wgpa19l6xar0k2m117iz2kq3cw433llp07sqswpf5969y698bf5";
			};

			beamDeps = [];
		};

		poison = buildMix rec {
			name = "poison";
			version = "3.1.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1kng8xadrs03i77irxvdk9vfncrqzncmgxc5gc8y8gkknw76dj7y";
			};

			beamDeps = [];
		};

		certifi = buildRebar3 rec {
			name = "certifi";
			version = "2.6.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0zmvagzisnk7lj5pfipl19mjq9wn70i339hpbkfljf0vk6s9fk2j";
			};

			beamDeps = [];
		};

		siphash = buildMix rec {
			name = "siphash";
			version = "3.2.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1hxpn31m9w1xyfw8za4v0wxpq24r9256w62yfix675bf3iq10f5s";
			};

			beamDeps = [];
		};

		phoenix = buildMix rec {
			name = "phoenix";
			version = "1.5.8";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0mfr691wb95ym4jj2jfrph0x5k1fha43phxnga61cdj85yix1pim";
			};

			beamDeps = [ jason phoenix_html phoenix_pubsub plug plug_cowboy plug_crypto telemetry ];
		};

		appsignal_plug = buildMix rec {
			name = "appsignal_plug";
			version = "2.0.4";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1b434bcq5lnrjmdjin9nqg5sxi9b3na0x5cn7i0j5gi8dwgb0i56";
			};

			beamDeps = [ appsignal plug ];
		};

		telemetry = buildRebar3 rec {
			name = "telemetry";
			version = "0.4.3";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0hc0fr2bh97wah9ycpm7hb5jdqr5hnl1s3b2ibbbx9gxbwvbhwpb";
			};

			beamDeps = [];
		};

		erlex = buildMix rec {
			name = "erlex";
			version = "0.2.6";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0x8c1j62y748ldvlh46sxzv5514rpzm809vxn594vd7y25by5lif";
			};

			beamDeps = [];
		};

		nanoid = buildMix rec {
			name = "nanoid";
			version = "2.0.5";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0n2z7pjkg2ksivn56ryj9is6pcr6zwwhaxs8m9rdl10i69v8hvlm";
			};

			beamDeps = [];
		};

		cowboy_telemetry = buildRebar3 rec {
			name = "cowboy_telemetry";
			version = "0.3.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1bzhcdq12p837cii2jgvzjyrffiwgm5bsb1pra2an3hkcqrzsvis";
			};

			beamDeps = [ cowboy telemetry ];
		};

		geolix = buildMix rec {
			name = "geolix";
			version = "1.1.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0033n5zzd4lsvrsippjzab209mq6g0kcb1lyqy6jh37kmvr5824q";
			};

			beamDeps = [ poolboy ];
		};

		postgrex = buildMix rec {
			name = "postgrex";
			version = "0.15.9";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1vmd63vxwz8knid424b0rbp200vj7q7rz3xp98yj5cjc7q81j1v1";
			};

			beamDeps = [ connection db_connection decimal jason ];
		};

		parallel_stream = buildMix rec {
			name = "parallel_stream";
			version = "1.0.6";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1sr3127izwy40j6sr25j745if70nblravwj2xfwqf6z1963jx6v3";
			};

			beamDeps = [];
		};

		excoveralls = buildMix rec {
			name = "excoveralls";
			version = "0.13.4";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "14313qwlvcj4yv3sglhph9r45ycigjd6c2y1881xyp73xsrh1bps";
			};

			beamDeps = [ hackney jason ];
		};

		php_serializer = buildMix rec {
			name = "php_serializer";
			version = "0.9.2";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "16m5ghdb74mxrnf9w31rq25cyz9fq1iv6wv747ygfi098rd87srl";
			};

			beamDeps = [];
		};

		bamboo_phoenix = buildMix rec {
			name = "bamboo_phoenix";
			version = "1.0.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1vkxnfw4xmaz7339nmbcrqlpr260g7c2pfwlg6j897014sxqzf3d";
			};

			beamDeps = [ bamboo phoenix ];
		};

		jumper = buildMix rec {
			name = "jumper";
			version = "1.0.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0cvlbfkapkvbwaijmjq3cxg5m6yv4rh69wvss9kfj862i83mk31i";
			};

			beamDeps = [];
		};

		junit_formatter = buildMix rec {
			name = "junit_formatter";
			version = "3.1.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1x3j7gbv2sk0mvjgrxcnbnnzq5lsiykdvazzfx7zq4gpjcd40lns";
			};

			beamDeps = [];
		};

		csv = buildMix rec {
			name = "csv";
			version = "2.4.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0ychghqb3smkxy42dz6pcpb9anddwc39dx0fn5k7kqk7mhw8jl2l";
			};

			beamDeps = [ parallel_stream ];
		};

		httpoison = buildMix rec {
			name = "httpoison";
			version = "1.8.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0fiwkdrbj7mmz449skp7laz2jdwsqn3svddncmicd46gk2m9w218";
			};

			beamDeps = [ hackney ];
		};

		envy = buildMix rec {
			name = "envy";
			version = "1.1.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "129s5hjw957i12qmq9k094s4h7hbvh8fr3ax2ibxfps18wdfnqbh";
			};

			beamDeps = [];
		};

		decimal = buildMix rec {
			name = "decimal";
			version = "1.9.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "11bxdzywr4pam7l0jpfrknhzlnz9zpgz473m7s7r5mpfd0sk9wmi";
			};

			beamDeps = [];
		};

		db_connection = buildMix rec {
			name = "db_connection";
			version = "2.2.2";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0307z0v1awc3scjvh955w4kw7gdw5g6nynmsnj9zpbx8v10g4ak4";
			};

			beamDeps = [ connection ];
		};

		idna = buildRebar3 rec {
			name = "idna";
			version = "6.1.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1sjcjibl34sprpf1dgdmzfww24xlyy34lpj7mhcys4j4i6vnwdwj";
			};

			beamDeps = [ unicode_util_compat ];
		};

		sshex = buildMix rec {
			name = "sshex";
			version = "2.2.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1nqkgf4ymcmgwzrvdsh3p6yfi8vq4svf6xrd1bkm1j0x06jwmcj5";
			};

			beamDeps = [];
		};

		credo = buildMix rec {
			name = "credo";
			version = "1.5.4";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1i0v2ivh4m8kvh9kllxpyws1pj8mqkd6zd8kp8wky2nwx92sylfg";
			};

			beamDeps = [ bunt file_system jason ];
		};

		decorator = buildMix rec {
			name = "decorator";
			version = "1.3.2";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1z57s5x4f2d24kj32lzj9bcs15qmgcqfskm8krnrwmy8wf4x02xq";
			};

			beamDeps = [];
		};

		parse_trans = buildRebar3 rec {
			name = "parse_trans";
			version = "3.3.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "12w8ai6b5s6b4hnvkav7hwxd846zdd74r32f84nkcmjzi1vrbk87";
			};

			beamDeps = [];
		};

		cachex = buildMix rec {
			name = "cachex";
			version = "3.3.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "189irin4xkbnj6b3ih1h5fvli1xq6m1sz1xiyqryyk71vphmw3nr";
			};

			beamDeps = [ eternal jumper sleeplocks unsafe ];
		};

		kaffy = buildMix rec {
			name = "kaffy";
			version = "0.9.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1lvf9h6pvzyyyn53axaf0g3snmyk1i8p27nimqrv9zk8irxzb3yi";
			};

			beamDeps = [ ecto phoenix phoenix_html ];
		};

		tzdata = buildMix rec {
			name = "tzdata";
			version = "1.0.5";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0nia83zpk0pb4jkpvhkmmgw8i5p6kd6cf776q6aj0pcym6i9llam";
			};

			beamDeps = [ hackney ];
		};

		elixir_uuid = buildMix rec {
			name = "elixir_uuid";
			version = "1.2.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0ljpb1l9v262k043irj6i6w9fwyqn0b2fj86jyhcwm9mdkma5szp";
			};

			beamDeps = [];
		};

		jason = buildMix rec {
			name = "jason";
			version = "1.1.2";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1zispkj3s923izkwkj2xvaxicd7m0vi2xnhnvvhkl82qm2y47y7x";
			};

			beamDeps = [ decimal ];
		};

		dialyxir = buildMix rec {
			name = "dialyxir";
			version = "1.0.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0m26cpw56gq1q47x1w7p5jy3sxaj5la1l1nq13519b2z2j46bc5f";
			};

			beamDeps = [ erlex ];
		};

		timex = buildMix rec {
			name = "timex";
			version = "3.6.3";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1v6rn17llandjdm8m3gdc2ylardsqn9czcww58888mngbzwz8sbd";
			};

			beamDeps = [ combine gettext tzdata ];
		};

		mime = buildMix rec {
			name = "mime";
			version = "1.6.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "19qrpnmaf3w8bblvkv6z5g82hzd10rhc7bqxvqyi88c37xhsi89i";
			};

			beamDeps = [];
		};

		unicode_util_compat = buildRebar3 rec {
			name = "unicode_util_compat";
			version = "0.7.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "08952lw8cjdw8w171lv8wqbrxc4rcmb3jhkrdb7n06gngpbfdvi5";
			};

			beamDeps = [];
		};

		cowlib = buildRebar3 rec {
			name = "cowlib";
			version = "2.9.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0w318ynls1iqvl5p8icyfbhysf12qvhr220narhrj3d78315s5z4";
			};

			beamDeps = [];
		};

		sentry = buildMix rec {
			name = "sentry";
			version = "8.0.5";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0z2bpm4j0x5liyw3d4575x7pc7s2vy54qs9ygdnqhbpmvfgq6wj9";
			};

			beamDeps = [ hackney jason plug plug_cowboy ];
		};

		phoenix_pagination = buildMix rec {
			name = "phoenix_pagination";
			version = "0.7.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0zw2dlfnfqgk0igm8dgldf8zp8m9f94y2whdpc1b6jw2x4w56197";
			};

			beamDeps = [ ecto_sql phoenix_html plug ];
		};

		comeonin = buildMix rec {
			name = "comeonin";
			version = "5.3.2";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "012zr4s7b5bipng6yszqxkqr1lcv7imf8gyvxad56jachh1396fh";
			};

			beamDeps = [];
		};

		yamerl = buildRebar3 rec {
			name = "yamerl";
			version = "0.8.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0gxjrrdrz9rcfmkg7x7z9p0fcz90dzqyk4laxz8gwi23svwk1jwn";
			};

			beamDeps = [];
		};

		ex_machina = buildMix rec {
			name = "ex_machina";
			version = "2.5.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1160arhhvvj0pk2dh8gs1lp6m9qyh3jg2d8kn6nziydk68f8a94g";
			};

			beamDeps = [ ecto ecto_sql ];
		};

		elixir_make = buildMix rec {
			name = "elixir_make";
			version = "0.6.2";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1na8agkks1hrwq1lxfj4yd96bvfcs4hk7mbra9z6lli2vanrxr03";
			};

			beamDeps = [];
		};

		double = buildMix rec {
			name = "double";
			version = "0.7.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0abk4nxq3bkxjkm63c6a5a5yy6n1xs7mj0xbggd54i3b4si8ghzh";
			};

			beamDeps = [];
		};

		cowboy = buildErlangMk rec {
			name = "cowboy";
			version = "2.8.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "12mcbyqyxjynzldcfm7kbpxb7l7swqyq0x9c2m6nvjaalzxy8hs6";
			};

			beamDeps = [ cowlib ranch ];
		};

		hackney = buildRebar3 rec {
			name = "hackney";
			version = "1.17.4";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "05kbk3rpw2j3cb9pybikydxmi2nm5pidpx0jsm48av2mjr4zy5ny";
			};

			beamDeps = [ certifi idna metrics mimerl parse_trans ssl_verify_fun unicode_util_compat ];
		};

		ref_inspector = buildMix rec {
			name = "ref_inspector";
			version = "1.3.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0m1q4fdy79hqrh7fp9a3l5pm8mqjd47gxqwnyxk9kip510dynwii";
			};

			beamDeps = [ hackney yamerl ];
		};

		gen_smtp = buildRebar3 rec {
			name = "gen_smtp";
			version = "1.1.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1dwszws49n7sza38js2999yzl3mn1yirxhyb9114mzby07651g2i";
			};

			beamDeps = [ hut ranch ];
		};

		metrics = buildRebar3 rec {
			name = "metrics";
			version = "1.0.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "05lz15piphyhvvm3d1ldjyw0zsrvz50d2m5f2q3s8x2gvkfrmc39";
			};

			beamDeps = [];
		};

		geolix_adapter_mmdb2 = buildMix rec {
			name = "geolix_adapter_mmdb2";
			version = "0.5.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0gndrrdr5paz1vgs0lw44pqxn62p4j5441wjji0yblx2l2v8a56b";
			};

			beamDeps = [ geolix mmdb2_decoder ];
		};

		sleeplocks = buildRebar3 rec {
			name = "sleeplocks";
			version = "1.1.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1q823i5bisc83pyssgrqkggyxiasm7b8dygzj2r943adzyp3gvl4";
			};

			beamDeps = [];
		};

		bcrypt_elixir = buildMix rec {
			name = "bcrypt_elixir";
			version = "2.3.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0s9psinz913l690xbcrl21m23zwinw4r2ypjgg7ybl3f9wfxd09c";
			};

			beamDeps = [ comeonin elixir_make ];
		};

		file_system = buildMix rec {
			name = "file_system";
			version = "0.2.10";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1p0myxmnjjds8bbg69dd6fvhk8q3n7lb78zd4qvmjajnzgdmw6a1";
			};

			beamDeps = [];
		};

		ua_inspector = buildMix rec {
			name = "ua_inspector";
			version = "0.20.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "09kz0bb3mj9w3mq8khlj870jwyg81gaszhigw65xbrsmkwxn5s1h";
			};

			beamDeps = [ hackney poolboy yamerl ];
		};

		combination = buildMix rec {
			name = "combination";
			version = "0.0.3";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1jhqvaggg55qvrlkz182b5n7pdbhf062a79pqryyyhnzcgs9kc3j";
			};

			beamDeps = [];
		};

		plug_crypto = buildMix rec {
			name = "plug_crypto";
			version = "1.2.2";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1nxnxj62iv4yvm4771jbxpj3l4brn2crz053y12s998lv5x1qqw7";
			};

			beamDeps = [];
		};

		cors_plug = buildMix rec {
			name = "cors_plug";
			version = "1.5.2";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0sfc0a5n8r4przz0pspyf5c9cqhcgqj8fsss6k2d0bf11p92gw4s";
			};

			beamDeps = [ plug ];
		};

		hut = buildRebar3 rec {
			name = "hut";
			version = "1.3.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0qxmkazkakrmvd9n1im4ad43wxgh189fqcd9lfsz58fqan2x45by";
			};

			beamDeps = [];
		};

		gettext = buildMix rec {
			name = "gettext";
			version = "0.18.2";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1igmn69xzj5wpkblg3k9v7wa2fjc2j0cncwx0grk1pag7nqkgxgr";
			};

			beamDeps = [];
		};

		poolboy = buildRebar3 rec {
			name = "poolboy";
			version = "1.5.2";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1qq116314418jp4skxg8c6jx29fwp688a738lgaz6h2lrq29gmys";
			};

			beamDeps = [];
		};

		connection = buildMix rec {
			name = "connection";
			version = "1.0.4";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1ng420jq6f0rk8k417i0bkaggxaiq0bsnw8ajbwkm912pv4m022a";
			};

			beamDeps = [];
		};

		bamboo = buildMix rec {
			name = "bamboo";
			version = "2.0.2";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0533pzfhqkh1haivc70s2pxm92f88gmk49x7790ikb6d9z7mg385";
			};

			beamDeps = [ hackney jason mime plug ];
		};

		unsafe = buildMix rec {
			name = "unsafe";
			version = "1.0.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1rahpgz1lsd66r7ycns1ryz2qymamz1anrlps986900lsai2jxvc";
			};

			beamDeps = [];
		};

		ranch = buildRebar3 rec {
			name = "ranch";
			version = "1.7.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1my7mz3x7a1fmjyin55nn1fr2d2rl3y64qf3kpcidxvxg0kqa7a5";
			};

			beamDeps = [];
		};

		appsignal_phoenix = buildMix rec {
			name = "appsignal_phoenix";
			version = "2.0.3";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "07wif63y62vh0i0vpyvisd6pya33ygb0qfgpmgngfwlv997yksii";
			};

			beamDeps = [ appsignal_plug phoenix phoenix_html telemetry ];
		};

		combine = buildMix rec {
			name = "combine";
			version = "0.10.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "06s5y8b0snr1s5ax9v3s7rc6c8xf5vj6878d1mc7cc07j0bvq78v";
			};

			beamDeps = [];
		};

		oban = buildMix rec {
			name = "oban";
			version = "2.6.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1sjllaa0ir1rwa6hdclyf2b9dlaj8fnldzbw9zv1l76xiymym1iy";
			};

			beamDeps = [ ecto_sql jason postgrex telemetry ];
		};

		mix_test_watch = buildMix rec {
			name = "mix_test_watch";
			version = "1.0.2";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "01qs8kfbzz0rv9k7dsvsp6yhln8mri7x1ikj75vq9xh6if6mbb27";
			};

			beamDeps = [ file_system ];
		};

		phoenix_live_reload = buildMix rec {
			name = "phoenix_live_reload";
			version = "1.3.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "10mmjmpp278kvsmqnrw4g63pjwb13sdqc6hwqrpbgwlkwdx01h82";
			};

			beamDeps = [ file_system phoenix ];
		};

		browser = buildMix rec {
			name = "browser";
			version = "0.4.4";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "1mzqniwx1px25mg79y7nq4iv7g5b18wq00w75ds1jjsaklqclxnl";
			};

			beamDeps = [ plug ];
		};

		plug = buildMix rec {
			name = "plug";
			version = "1.11.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "07w83cx4xx90x4l1kmil4lpby55lpw83jfw3y08pqn5vxx7lwli3";
			};

			beamDeps = [ mime plug_crypto telemetry ];
		};

		bertex = buildMix rec {
			name = "bertex";
			version = "1.3.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0yp5iqyj477yl8hz3788igjh04j9rbhwldxfgdxlnxmmid3mwp8a";
			};

			beamDeps = [];
		};

		appsignal = buildMix rec {
			name = "appsignal";
			version = "2.0.8";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "092s07aj4ad81iybpc7m1sn3mvg2wqjdnisyyxrmhvp27qzrx7yi";
			};

			beamDeps = [ decorator hackney jason poison telemetry ];
		};

		mimerl = buildRebar3 rec {
			name = "mimerl";
			version = "1.2.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "08wkw73dy449n68ssrkz57gikfzqk3vfnf264s31jn5aa1b5hy7j";
			};

			beamDeps = [];
		};

		bamboo_smtp = buildMix rec {
			name = "bamboo_smtp";
			version = "4.0.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0ggjm05yf8psvvw98fijzfd5yp4fv6y59rpjycj9y6qjj98024i4";
			};

			beamDeps = [ bamboo gen_smtp ];
		};

		bunt = buildMix rec {
			name = "bunt";
			version = "0.2.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0lw3v9kwbbcy1v6ygziiky887gffwwmxvyg4r1v0zm71kzhcgxbs";
			};

			beamDeps = [];
		};

		plug_cowboy = buildMix rec {
			name = "plug_cowboy";
			version = "2.4.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "01q5ny2m75206987xgzjaakskqq8524nhnrakdypmczpvyv168fp";
			};

			beamDeps = [ cowboy cowboy_telemetry plug telemetry ];
		};

		mmdb2_decoder = buildMix rec {
			name = "mmdb2_decoder";
			version = "3.0.0";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "13pa1gr8c7pdw8i9vr32rrvcl790f9ndkxmrrqfqslqm54jck79m";
			};

			beamDeps = [];
		};

		ecto = buildMix rec {
			name = "ecto";
			version = "3.6.1";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "0lv5mypffqaxqkzhv78d9qvhdy4ckxs8lj150ygv2iq4k552kcyb";
			};

			beamDeps = [ decimal jason telemetry ];
		};

		eternal = buildMix rec {
			name = "eternal";
			version = "1.2.2";

			src = fetchHex {
				pkg = "${name}";
				version = "${version}";
				sha256 = "10p7m6kv2z2c16gw36wgiwnkykss4lfkmm71llxp09ipkhmy77rc";
			};

			beamDeps = [];
		};
	};
in self

```

happysalada replied:
Just a heads up regarding mixDeps. I've merged the PR integrating nix dependencies into mixRelease. If you want to use nix dependencies the attribute you have to pass is mixNixDeps. You can continue to use the fod dependencies, you'll just have to change the name from `mixDeps` to `mixFodDeps`

Ma27 replied:
With `mixNIxDeps` generated & set I get the following compilation error:

```
plausible> == Compilation error in file lib/plausible/clickhouse_repo.ex ==
plausible> ** (ArgumentError) adapter ClickhouseEcto was not compiled, ensure it is correct and it is included as a project dependency
plausible>     (ecto 3.6.1) lib/ecto/repo/supervisor.ex:60: Ecto.Repo.Supervisor.compile_config/2
plausible>     lib/plausible/clickhouse_repo.ex:2: (module)
plausible>     (stdlib 3.14.2) erl_eval.erl:680: :erl_eval.do_apply/6
error: build of '/nix/store/283178iklq7nsbr9dqr443a0lklsjs4h-plausible-1.3.0.drv' on 'ssh-ng://builder-ng' failed: error: builder for '/nix/store/283178iklq7nsbr9dqr443a0lklsjs4h-plausible-1.3.0.drv' failed with exit code 1
error: builder for '/nix/store/283178iklq7nsbr9dqr443a0lklsjs4h-plausible-1.3.0.drv' failed with exit code 1

```

An idea what's going on here?

happysalada replied:
Ah, the clickhouse dependency are git dependencies. mix2nix currently doesn't support git dependencendies. That's why you are seeing this missing dependency.
You'll have to use FOD until we figure out how to do the git dependencies properly

Ma27 replied:
Ok, thanks for the clarification! I'll resolve this for now as there's nothing actionable atm :)

happysalada replied:
you just need to remember to change mixDeps to `mixFodDeps` that's it.

happysalada commented on pkgs/servers/web-apps/plausible/default.nix:
```diff
@@ -0,0 +1,85 @@
+    mix deps.compile --path $out --no-deps-check
+    mix compile --no-deps-check --path $out
+
+    npm run deploy --prefix ./assets
```
the mix digest task is missing. On some applications it's not required. Just mentioning it here in case you forgot.

Ma27 replied:
Yeah, should've added a comment in the first place, sorry!
The thing is, when I run `mix phx.digest`, it was complaining about missing dependencies and I didn't really understand why. Do you have an idea or want to take a look? Otherwise I can do it next week (IIRC this is used for minify tasks and such, I tested plausible a bit and it seemed to run OK for me, so maybe there's no actual need anyways).

happysalada replied:
you need to run it with the `--no-deps-check` flag. Let me know if you still have a problem and I'll take a look.

happysalada replied:
The digest task is more for cache invalidation. If you don't run it, when an end user loads the website, there is a chance that it will just use the cached version. Digest gives a unique name that will make sure that a new version is downloaded when there is a need.

Ma27 replied:
Hmm would you mind taking a look? Even with `--no-deps-check` it errors out because of dependency issues.

happysalada commented on pkgs/servers/web-apps/plausible/default.nix:
```diff
@@ -0,0 +1,85 @@
+
+    mkdir -p $out
+    ln -sf ${yarnDeps}/node_modules assets/node_modules
+    mix deps.compile --path $out --no-deps-check
```
deps.compile should already be done in the configure phase.

I'm not sure you need to override the whole buildPhase here.

Probably just adding a `postConfigure` steps with the yarn stuff should be enough
something like
```
ln -sf ${yarnDeps}/node_modules assets/node_modules
npm run deploy --prefix ./assets
mix phx.digest --no-deps-check
```

Ma27 replied:
> I'm not sure you need to override the whole buildPhase here.

Tbh, I mostly redid the steps from the `Dockerfile` (and IIRC `mixRelease` didn't exist when I wrote a draft in Februrary). But yeah, I should've taken a closer look here, sorry! Will do next week.

But `buildPhase` (or `postBuild` - idk) is at least needed for the `npm` stuff.

happysalada replied:
yes, to run phx.digest you need postBuild, you're right.

happysalada commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,270 @@
+            ++ optional cfg.database.postgres.setup config.services.postgresql.package;
+
+          serviceConfig = {
+            DynamicUser = true;
```
Just curious to discuss a little this.
The default postgres authentication going through the socket will be `peer`, not having a fixed user run the app, that would mean that you have to change the authentication mechanism to `trust` right?

The alternative being to create a user `plausible` and run it from him. You can make it a `isSystemUser`. 

Just want to make sure I'm not missing something here.

Ma27 replied:
The `DynamicUser` is called `plausible` (i.e. equal to the service-name) as long as the unit is running. Since `plausible` is also a valid postgresql user, the peer auth works fine, so AFAIU no need to statically add a user.

happysalada commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,270 @@
+            ''} plausible";
+          };
+        };
+      }
```
Would there be a need to add a service for a remote shell ?
In order to be able to connect to a running instance, you need to `/bin/plausible remote` with the necessary environment variable.
I'm just thinking it might be a very nice convenience in order to debug or check what is happening. It definitely might not be useful for the average user.

Ma27 replied:
Good idea! I guess it's nicer though to have a wrapper-script for that which can be executed from the shell. An example for what I mean would be `nextcloud-occ`.

happysalada replied:
It looks good! I'm not sure I understand the way secrets are managed, but otherwise it looks good!

Ma27 replied:
Hmm even if the environment is correctly set up, `plausible remote` complains about a missing HOME dir even though it's set (checked by looking at `/proc/pid/environ`)... will take a look at this later.

aanderse commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,270 @@
+                fi
+              ''}
+            ''} plausible-setup";
+            ExecStart = "@${pkgs.writeShellScript "plausible" ''
```
Almost enough to make you want to simplify and use `EnvironmentFile` to back each of the secrets options...

Ma27 replied:
IIRC `EnvironmentFile` doesn't allow file-substitutions such as `$(<...)` and I thought that having `xxxFile`-options is generally nicer because it can be trivially integrated with e.g. `sops-nix` or NixOps' `deployment.keys` feature.

aanderse replied:
I've noticed the pattern that people ask for both the key and value in secrets files.

You still could have an option per file, but simply append each option to the `EnvironmentFile` list.

I'm mostly just mentioning this as an alternative to avoid shell scripting. Both ways have pros and cons.

happysalada replied:
Just as an additional data point, I use EnvironmentFile with agenix (incidentally on an elixir service), and the integration works really well.

On a separate note, systemd doesn't recommend environment variable or EnvironmentFile for secrets. I've been trying to make LoadCredentials work (what they actually recommend) without success.

aanderse replied:
I'm aware they don't recommend `EnvironmentFile` for secrets... but they don't have a world readable nix store to deal with 😄 

You're right, though. I should have mentioned `LoadCredential`, especially because this runs as `DynamicUser`, which implies these secrets are already world readable the way this module is.

Great suggestion!

happysalada replied:
There is an ongoing conversation regarding LoadCredential at the moment with the systemd guys.
It is broken with the current systemd (247) but it should be fixed with the next one. There is a PR for it at the moment but not merged.
Second point is you can only use files for credentials. I'm still a bit confused how this should be used with environment variable for example. Should you separate all your env vars into separate files and then source them in your web application? I'm not sure.
Trying to give more information in case you are considering LoadCredential. Not making a recommendation for it though.

Ma27 replied:
>  Second point is you can only use files for credentials. I'm still a bit confused how this should be used with environment variable for example. Should you separate all your env vars into separate files and then source them in your web application? I'm not sure.

Perhaps something like `export FOO=$(<secret)`? Basically what I'm doing now except that you have to manage these files somehow else (e.g. with sops-nix or `deployment.keys`).

happysalada replied:
Thanks for the recommendation, I'm asking them directly, will let you know if I find out anything.

aanderse reviewed (commented):
Module LGTM and I approve as is, though as mentioned I think passing a few files to `EnvironmentFile` instead of the accepting situation might be beneficial... but hey that's just my opinion.

I don't know anything about `elixir` packaging so maybe someone else can take a look at the package portion of this PR.

Ma27 commented:
Thanks a lot for the reviews folks! Will work on all that next week :)

happysalada commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,270 @@
+  envSecrets = ''
+    export ADMIN_USER_PWD="$(<${cfg.adminUser.passwordFile})"
+    export SECRET_KEY_BASE="$(<${cfg.server.secretKeybaseFile})"
+    export RELEASE_TMP=/var/lib/plausible/tmp
```
not sure this one needs to be a secret. You could potentially just add it to the environment

happysalada commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,270 @@
+          default = "http://localhost:8123/default";
+          type = types.str;
+          description = ''
+            The URL to be used to connect to <package>postgres</package>. The format
```
perhaps you meant clickhouse here?

happysalada commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,270 @@
+    };
+
+    services.clickhouse = mkIf cfg.database.clickhouse.setup {
+      enable = true;
```
I was checking and the nixpkgs version of clickhouse is 6 months old.
I remember on the plausible project, there was a discussion about pinning the clickhouse version in the docker thing. We should probably just make sure that the clickhouse db is a new enough version at least.

Ma27 replied:
Do you think it's worth the effort? As in: how likely are breaking changes that'll bite us? I'm wondering if it's OK to leave it as-is (we have a VM test that should help us find these issues) and whether we should make these changes in case they become necessary.

happysalada replied:
I really don't know, we can probably leave it as is for now.

lheckemann commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,270 @@
+    };
+
+    server = {
+      disableRegistration = mkOption {
```
Is this the upstream option naming? I generally prefer positive names (e.g. `allowRegistration` or `enableRegistration`) because less negation means less cognitive overhead.

Ma27 replied:
I'm all for positivity as well and also disagree a bit with their naming (including the remark you've had) :)).

However I'm generally a fan of maintaining consistency with upstream if possible and given that we also have things like RFC42 now, this seems to be a general trend (note that I've decided against that here because I haven't found a nice way to transform the options into environment vars while properly securing secrets and there aren't too many options so I guess that's fine).

lheckemann commented on pkgs/servers/web-apps/plausible/update.sh:
```diff
@@ -0,0 +1,68 @@
+fixed_tailwind_version="$(jq '.dependencies.tailwindcss' -r <<< "$package_json" | sed -e 's,^^,,g')"
+
+echo "$package_json" \
+  | jq '. + {"name":"plausible","version":"'$latest'"}' \
+  | jq '.dependencies.tailwindcss = "'"$fixed_tailwind_version"'"' \
```
```suggestion
  | jq '. + {"name":"plausible","version": $ENV.latest} | .dependencies.tailwindcss = $ENV.fixed_tailwind_version' \
```

lheckemann commented on pkgs/servers/web-apps/plausible/update.sh:
```diff
@@ -0,0 +1,68 @@
+set -x
+
+dir="$(realpath $(dirname "$0"))"
+latest="$(curl -q https://api.github.com/repos/plausible/analytics/releases/latest \
```
```suggestion
export latest="$(curl -q https://api.github.com/repos/plausible/analytics/releases/latest \
```
(for later use)

lheckemann commented on pkgs/servers/web-apps/plausible/update.sh:
```diff
@@ -0,0 +1,68 @@
+SRC="https://raw.githubusercontent.com/plausible/analytics/${latest}"
+
+package_json="$(curl -qf "$SRC/assets/package.json")"
+fixed_tailwind_version="$(jq '.dependencies.tailwindcss' -r <<< "$package_json" | sed -e 's,^^,,g')"
```
```suggestion
export fixed_tailwind_version="$(jq '.dependencies.tailwindcss' -r <<< "$package_json" | sed -e 's,^^,,g')"
```
for later use

lheckemann commented on pkgs/servers/web-apps/plausible/update.sh:
```diff
@@ -0,0 +1,68 @@
+  | sed -e 's,../deps/,../../tmp/deps/,g' \
+  > $dir/package.json
+
+tarball_meta="$(nix-prefetch-git https://github.com/plausible/analytics --rev "$latest" --quiet)"
```
This seems somewhat error-prone: I wouldn't rely on `nix-prefetch-git` producing the same output as the tarball fetched by `fetchFromGitHub` — and it's very easy to miss if this does happen when updating, because of course it'll build successfully on the system where nix-prefetch-git ran! I'd prefer `nix-prefetch-url --unpack https://github.com/plausible/analytics/archive/"$latest".tar.gz --print-path` to get the hash and path.

Ma27 replied:
Using `nix-prefetch-github` now :)

lheckemann commented on pkgs/servers/web-apps/plausible/update.sh:
```diff
@@ -0,0 +1,68 @@
+
+cp -r $tarball_path/* $tmp_setup_dir/
+cp -r "$(nix-build -A plausible.mixDeps)" "$tmp_setup_dir/deps"
+chmod -R a+rwx "$tmp_setup_dir"
```
```suggestion
chmod -R u+rwX "$tmp_setup_dir"
```
or does anything actually need to be accessible by the user and do any of the files need to be made executable?

lheckemann reviewed (changes requested):
Nice work!

Ma27 commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,270 @@
+                /run/wrappers/bin/sudo -Hu postgres ${config.services.postgresql.package}/bin/psql --port=5432 "$@"
+              }
+              PSQL -tAc "CREATE EXTENSION IF NOT EXISTS citext;"
+              PSQL -tAc "CREATE ROLE plausible WITH LOGIN;"
+              PSQL -tAc "CREATE DATABASE plausible WITH OWNER plausible;"
+              PSQL -tAc "ALTER USER plausible WITH SUPERUSER;"
```
I'll check what exactly this is for, but plausible refuses to start without that, so no matter what I'll find it will make some more patching necessary.

happysalada replied:
I'm also curious about why super user priviledge is required. Just curious though. What is the error when it's not superuser priviledge?

Ma27 replied:
Hmm weird, it's an "insufficient privilege" because of a missing right to create a database. I guess I gotta find out why it still wants to create it :/

happysalada replied:
in the migrations, there is probably a "create extention". You need to be superuser to do those.
(I ran into the same problem with another app using postgis for example)
You can probably get away without needing superuser priviledge if you ensure that the extension is there.

Ma27 replied:
patched away the superuser requirement :)

SuperSandro2000 commented on nixos/modules/services/web-apps/plausible.xml:
```diff
@@ -0,0 +1,51 @@
+<chapter xmlns="http://docbook.org/ns/docbook"
```
Please convert nixos docs to markdown.

SuperSandro2000 commented on nixos/tests/plausible.nix:
```diff
@@ -0,0 +1,46 @@
+import ./make-test-python.nix ({ pkgs, lib, ... }: {
+  name = "plausible";
+  meta = with pkgs.stdenv.lib.maintainers; {
```
```suggestion
  meta = with lib.maintainers; {
```

If we already pass in lib we can use it to make the variables shorter.

Ma27 committed:
beam/fetchMixDeps: allow adding patches

Needed to modify e.g. a few things in the `plausible` package to get it
running properly.

Ma27 committed:
plausible: first review fix iteration

* Most significant is probably the patching necessary to run plausible
  with postgres without superuser privilege. This change includes:
  * updating ecto_sql to 3.6 where `CREATE DATABASE` is only executed if
    it doesn't exist[1].
  * patching a migration to only modify the `users.email` column (to use
    `citext` rather than creating the extension. `plausible-postgres`
    takes care of that).
* Correctly declare dependencies in systemd.
* A few minor fixes.

[1] https://github.com/elixir-ecto/ecto_sql/commit/051baf669ef6ec92639ccc8b113d00b98047e927

Ma27 commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,270 @@
+          partOf = [ "plausible.service" ];
+          serviceConfig.Type = "oneshot";
+          script = ''
+            if [ ! -e /var/lib/plausible/.db-setup ]; then
```
Done!

Ma27 committed:
plausiblew: cleanup build & update script

happysalada commented on pkgs/servers/web-apps/plausible/default.nix:
```diff
@@ -0,0 +1,88 @@
+  postPatch = ''
+    # Without this modification, tzdata tries to write in its store-path:
+    # https://github.com/lau/tzdata#data-directory-and-releases
+    echo 'config :tzdata, :data_dir, (System.get_env("PLAUSIBLE_TZDATA") || "/tmp/plausible_tzdata")' \
```
I know it's already a lot of work, but how about upstreaming this.
You could have it like
`System.get_env("TZDATA_DIR", "priv")` (the default is priv)
I don't think they would complain.

happysalada replied:
All of my comments are suggestion, feel free to do only the ones for which you have energy for.

happysalada replied:
actually, https://github.com/plausible/analytics/pull/1096

happysalada commented on pkgs/servers/web-apps/plausible/default.nix:
```diff
@@ -0,0 +1,88 @@
+{ lib, stdenv, beamPackages, fetchFromGitHub, glibcLocales, cacert
```
really a nitpick, but would you mind running the nixpkgs-fmt formatter.
Just so there is consistent formatting

happysalada commented on pkgs/servers/web-apps/plausible/default.nix:
```diff
@@ -0,0 +1,88 @@
+      >> config/config.exs
+  '';
+
+  buildPhase = ''
+    runHook preBuild
+
+    mkdir -p $out
+    ln -sf ${yarnDeps}/node_modules assets/node_modules
+    npm run deploy --prefix ./assets
+    mix release plausible --no-deps-check --path $out
+
+    runHook postBuild
+  '';
```
```suggestion
  postBuild = ''
    ln -sf ${yarnDeps}/node_modules assets/node_modules
    npm run deploy --prefix ./assets
    # for external task you need a workaround for the no deps check flag
    # https://github.com/phoenixframework/phoenix/issues/2690
    mix do deps.loadpaths --no-deps-check, phx.digest
  '';
```

happysalada replied:
I realized I used npm to test, but you should be able  to replace with the equivalent `yarn` command if you prefer.

happysalada replied:
I would suggest to not override normal steps and just use the post step, if that's okay with you.

happysalada commented on pkgs/servers/web-apps/plausible/default.nix:
```diff
@@ -0,0 +1,88 @@
+    homepage = "https://plausible.io/";
+    description = " Simple, open-source, lightweight (< 1 KB) and privacy-friendly web analytics alternative to Google Analytics.";
+    maintainers = with maintainers; [ ma27 ];
+    platforms = platforms.linux;
```
```suggestion
    platforms = platforms.unix;
```
There should be no problem with darwin

happysalada commented on pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix:
```diff
@@ -62,7 +62,7 @@ in rec {
 
   mkYarnModules = {
-    name, # safe name and version, e.g. testcompany-one-modules-1.0.0
+    name ? "${pname}-${version}", # safe name and version, e.g. testcompany-one-modules-1.0.0
```
maybe that can be a separate commit. I think it can be merged quite quickly

Ma27 replied:
It is :)

happysalada replied:
Sorry, I meant separate PR. Since this PR is almost done, I don't have a problem with leaving as is.

happysalada commented on pkgs/servers/web-apps/plausible/default.nix:
```diff
@@ -0,0 +1,88 @@
+    patches = [ ./ecto_sql-fix.patch ];
+  };
+
+  yarnDeps = mkYarnModules {
```
Just curious here. Why did you choos yarn over node2nix for example?
I thought that the project was using yarn, but it's actually using npm by default.
If it's too much work, feel free to leave it as is. I'm thinking we should perhaps just add a comment as to why this choice was made in case future maintainers want to change this.

Ma27 replied:
From my experience `yarn2nix` is wayy more robust than `node2nix` so `yarn2nix` is usually my first choice :)

happysalada replied:
Maybe just add a comment saying that here yarn2nix was chosen over node2nix from the maintainers experience regarding stability.
The idea would be to make it clear to somebody in the future in case that decision needs to be revisited.

Ma27 committed:
plausible: correctly run digest task

Ma27 committed:
plausible: platforms.{linux -> unix}

Co-authored-by: Raphael Megzari <raphael@megzari.com>

Ma27 committed:
plausible: minor polishing

happysalada commented on nixos/tests/plausible.nix:
```diff
@@ -0,0 +1,46 @@
+      };
+      server = {
+        baseUrl = "http://localhost:8000";
+        secretKeybaseFile = "${pkgs.writeText "dont-try-this-at-home" "nannannannannannannannannannannannannannannannannannannan_batman!"}";
```
❤️

happysalada commented on pkgs/servers/web-apps/plausible/default.nix:
```diff
@@ -0,0 +1,92 @@
+
+    # CREATE EXTENSION requires super-user privileges. To avoid that, we just skip
+    # the responsible SQL statement here and take care of it in the module.
+    ./skip-create-ext.patch
```
I'm not sure you need this patch. If you create the extension before hand, it won't try to create it again.

Ma27 replied:
Seems as my mistake was that I added the CREATE EXTENSION before the CREATE DATABASE initially which made a second CREATE EXTENSION necessary in the migration script. I fixed it later, but already had the patch applied. It seems as the patch is actually not needed now, so I'll remove it :)

happysalada commented on pkgs/servers/web-apps/plausible/ecto_sql-fix.patch:
```diff
@@ -0,0 +1,13 @@
+diff --git a/mix.exs b/mix.exs
```
it looks like you're not applying this patch. Is it needed? If not perhaps we can remove it?

happysalada commented on pkgs/servers/web-apps/plausible/default.nix:
```diff
@@ -0,0 +1,92 @@
+    pname = "${pname}-deps";
+    inherit src version;
+    sha256 = "18h3hs69nw06msvs3nnymf6p94qd3x1f4d2zawqriy9fr5fz7zx6";
+    patches = [ ./ecto_sql-fix.patch ];
```
Perhaps just a comment on why the patch is needed?

happysalada reviewed (commented):
Mostly looks good to me.
A couple of comments added would be great but not an unbendable requirement.

What about the doc issue? I don't know anything about xml vs markdown, I'm willing to go either way, do we want to address the issue in this PR?

Ma27 committed:
plausible: remove create extension patch

Turns out this isn't needed if the module correctly adds the `citext`
extension to the `plausible` database.

happysalada commented:
Thanks a lot for doing this!

nh2 commented:
Hi, I'm currently trying this out.

How is one supposed to set `secretKeybaseFile` in production?

Because `DynamicUser` is used, I cannot know the user ID ahead of time that will need to read that file, and I certainly don't want to put it world-readable in the nix store.

Are there any pointers I could read about how one does that?

lheckemann commented:
Off the top of my head, I think `LoadCredential=` (see `systemd.exec(5)`) could be helpful for this, but I'm not sure if applying it here is trivial or not.

Ma27 commented:
I'm not sure right now, but IIRC it was sufficient to set e.g. `SupplementaryGroups` and make the secret readable by some sort of dedicated group (which is also the default approach used by e.g. deployment.keys of `nixops` or `sops-nix`). If you're having problems with this, I can check e.g. tomorrow :)

nh2 commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,273 @@
+
+  # FIXME consider using LoadCredential as soon as it actually works.
+  envSecrets = ''
+    export ADMIN_USER_PWD="$(<${cfg.adminUser.passwordFile})"
```
As pretty much 100% of shell scripts, this is probably wrong:

The shell this runs in doesn't have `set -e` nor any other reasonable way of error propagation, so when the configured file isn't readable by the `DynamicUser`, this code just prints e.g.

```
/nix/store/rwfwdk34fyz9qqyvn4vs0lnswscd73x0-plausible-setup: line 2: /var/lib/plausible/plausible-admin-user-password-file: Permission denied
/nix/store/rwfwdk34fyz9qqyvn4vs0lnswscd73x0-plausible-setup: line 3: /var/lib/plausible/plausible-secret-keybase-file: Permission denied
```

and then happily goes on with its day, with `ADMIN_USER_PWD` and `SECRET_KEY_BASE` set to the empty string.

nh2 replied:
Even if it had `set -e`, it would still be wrong, in the same way as the scripts of Elixir itself here:

https://github.com/elixir-lang/elixir/issues/11114

nh2 replied:
Fixes in PR https://github.com/NixOS/nixpkgs/pull/130062

nh2 commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,273 @@
+                  psql -d plausible <<< "UPDATE users SET email_verified=true;"
+                fi
+              ''}
+            ''} plausible-setup";
```
What is the argument to this shell script for?

It calls e.g.

```
/nix/store/skpa6a2l84ydidgxnpv63fnc1fsxyv6k-plausible-setup plausible-setup
```

But the shell script defined directly above does not use this argument.

Ma27 replied:
To quote `systemd.service(5)`:

> If the executable path is prefixed with "@", the second specified
> token will be passed as "argv[0]" to the executed process (instead of
> the actual filename), followed by the further arguments specified.

A similar approach for shorter names is done in e.g. the module of Hydra or the firewall :) 

Does that answer your question?

happysalada replied:
I actually wanted to ask you why you used @ in this case? (rather without)
You don't actually use the argument in this case right?

nh2 replied:
@Ma27 Ah yes, thank you! Is the intent to make it look better in `ps`/`htop`?

nh2 commented on nixos/modules/services/web-apps/plausible.nix:
```diff
@@ -0,0 +1,273 @@
+            ExecStart = "@${pkgs.writeShellScript "plausible" ''
+              ${envSecrets}
+              plausible start
+            ''} plausible";
```
Same question as https://github.com/NixOS/nixpkgs/pull/124055/files#r668400359

happysalada commented:
The "best" way to do this would be to use `LoadCredentials`.
The current problem is that systemd is blocked at version 147 and LoadCredentials doesn't work with that version.
The other problem is to make upstream be able to read a configuration variable from a file. I've worked with upstream for that change, and it should be available in the next release.

So in summary no perfect way to do this. I haven't found a way without LoadCredentials to not have your variable readable.

nh2 commented:
@happysalada Another problem I have with `DynamicUser` is that I cannot figure out how to debug things well with it.

For me the `plausible` startup fails with

```
** (RuntimeError) Code: 81, e.displayText() = DB::Exception: Database `` doesn't exist (version 20.5.2.7)
```

because in some of its calls from `createdb.sh`, it seems to not pass the `CLICKHOUSE_DATABASE_URL=http://127.0.0.1:8123/default` setting all the way through, and make `?database=` requests (with empty string as the DB name), right after making two `?database=default` requests.

The usual approach is to `sudo -u plausible /path/to/createdb.sh`, running it directly, perhaps with `strace`.

This seems impossible with `DynamicUser`, because there is no user you could `sudo -u` as.

So while I like the concept of `DynamicUser`, it seems to currently block these more fundamental debugging / sysadmin facilities.

Ma27 commented:
You should be able to do something like `systemd-run -p DynamicUser=yes -p StateDirectory=plausible -- your cmd`. If a `StateDirectory` is used, the DynamicUser will be named after and then you have an environment similar to the service itself.

happysalada commented:
I love this conversation!
I originally had some problems with DynamicUser for the backup. I ended up creating another User for backup and granting all priviledges on db for it.
I don't really like having to create another user and granting all priviledges. Is there some other black magic command from systemd that has something for that?
Or do you guys also have another user for backups on production?
(I use a systemd-timer, with a simple shell script that basically dumps the db, gathers other files like uploads and the likes and borgbackup it to a remote)

nh2 commented on pkgs/servers/web-apps/plausible/default.nix:
```diff
@@ -0,0 +1,92 @@
+
+    # We need ecto 3.6 as this version checks whether the database exists before
+    # trying to create it. The creation attempt would always require super-user privileges
+    # and since 3.6 this isn't the case anymore.
```
This patch breaks plausible for me:

* https://github.com/NixOS/nixpkgs/pull/130062#issuecomment-879559948
* https://github.com/NixOS/nixpkgs/pull/130062/commits/1656e8eb0de75cbfcc5006169a1ef4cf6140ca15
